File: app/src/main/java/br/com/onibusapp/onibusapp/ui/favoritos/FavoritosPresenter.java
package br.com.onibusapp.onibusapp.ui.favoritos;

import android.support.annotation.NonNull;

import java.util.List;

import br.com.onibusapp.onibusapp.R;
import br.com.onibusapp.onibusapp.adapter.RecyclerViewOnClickListener;
import br.com.onibusapp.onibusapp.data.dao.FavoritoDAO;
import br.com.onibusapp.onibusapp.data.dominio.Favorito;
import br.com.onibusapp.onibusapp.ui.MapsFragment;
import br.com.onibusapp.onibusapp.utils.Constantes;
import br.com.onibusapp.onibusapp.utils.FragmentUtil;

import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.FragmentManager;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by diego on 25/07/2018.
 */

public class FavoritosPresenter implements FavoritosContract.Presenter {

    private final FavoritoDAO favoritoDAO;
    private final FavoritosContract.View mFavoritosView;
    private List<Favorito> favoritos;
    private final RecyclerViewOnClickListener recyclerViewOnClickListener;
    private FragmentManager fragmentManager;

    public FavoritosPresenter(@NonNull FavoritosContract.View mFavoritosView,
                              @NonNull FavoritoDAO favoritoDAO,
                              @NonNull RecyclerViewOnClickListener recyclerViewOnClickListener,
                              @NonNull FragmentManager fragmentManager) {
        this.mFavoritosView = checkNotNull(mFavoritosView, "mFavoritosView cannot be null!");
        this.favoritoDAO = checkNotNull(favoritoDAO);
        this.recyclerViewOnClickListener = checkNotNull(recyclerViewOnClickListener);
        this.fragmentManager = fragmentManager;
        this.carregar();
    }

    private void carregar() {
        this.favoritos = this.favoritoDAO.findAll();
        mFavoritosView.carregar(this.favoritos);
    }

    @Override
    public void apagar(Integer posicao) {
        Favorito favorito = this.favoritos.get(posicao);
        this.favoritos.remove(posicao);
        this.favoritoDAO.apagar(favorito.getId());
        this.mFavoritosView.atualizar(posicao);
    }

    @Override
    public void carregarMapa(Integer posicao) {
        Favorito favorito = this.favoritos.get(posicao);
        FragmentUtil.getInstance(this.fragmentManager)
                    .criarBundle()
                    .parametros(Constantes.LINHA, favorito.getNomeLinha())
                    .parametros(Constantes.SENTIDO, favorito.getCodigoSentido())
                    .parametros(Constantes.CODIGO_EMPRESA, favorito.getCodigoEmpresa())
                    .mudarTela(new MapsFragment());
    }
}

File: app/src/main/java/br/com/onibusapp/onibusapp/ui/favoritos/FavoritosPresenter.kt
package br.com.onibusapp.onibusapp.ui.favoritos

import android.support.v4.app.FragmentManager
import br.com.onibusapp.onibusapp.adapter.RecyclerViewOnClickListener
import br.com.onibusapp.onibusapp.data.dao.FavoritoDAO
import br.com.onibusapp.onibusapp.data.dominio.Favorito
import br.com.onibusapp.onibusapp.ui.MapsFragment
import br.com.onibusapp.onibusapp.utils.Constantes
import br.com.onibusapp.onibusapp.utils.FragmentUtil

/**
 * Created by diego on 25/07/2018.
 */

class FavoritosPresenter(mFavoritosView: FavoritosContract.View,
                         favoritoDAO: FavoritoDAO,
                         recyclerViewOnClickListener: RecyclerViewOnClickListener,
                         private val fragmentManager: FragmentManager) : FavoritosContract.Presenter {

    private val favoritoDAO: FavoritoDAO
    private val mFavoritosView: FavoritosContract.View
    private var favoritos: MutableList<Favorito> = arrayListOf()
    private val recyclerViewOnClickListener: RecyclerViewOnClickListener

    init {
        this.mFavoritosView = kotlin.checkNotNull(mFavoritosView)
        this.favoritoDAO = kotlin.checkNotNull(favoritoDAO)
        this.recyclerViewOnClickListener = kotlin.checkNotNull(recyclerViewOnClickListener)
        this.carregar()
    }

    private fun carregar() {
        this.favoritos = this.favoritoDAO.findAll()
        mFavoritosView.carregar(this.favoritos!!)
    }

    override fun apagar(posicao: Int?) {
        val favorito = this.favoritos[posicao!!]
        //this.favoritos.removeAt(posicao)
        this.favoritoDAO.apagar(favorito.id)
        this.mFavoritosView.atualizar(posicao)
    }

    override fun carregarMapa(posicao: Int?) {
        val favorito = this.favoritos[posicao!!]
        FragmentUtil.getInstance(this.fragmentManager)
                .criarBundle()
                .parametros(Constantes.LINHA, favorito.linha)
                .parametros(Constantes.SENTIDO, favorito.codigoSentido)
                .parametros(Constantes.URL, favorito.url)
                .mudarTela(MapsFragment())
    }
}
